File: src/scanner.ts

import {
  SyntaxKind,
  Token,
  Identifier,
  StringLiteral,
  EndOfFile,
  BlockStart,
  BlockEnd,
  LineFoldEnd,
  PubKeyword,
  MutKeyword,
  LetKeyword,
  ImportKeyword,
  TypeKeyword,
  TextPosition,
  Colon,
  Comma,
  Equals,
  LParen,
  RParen,
  LBrace,
  LBracket,
  RBrace,
  RBracket,
  ReturnKeyword,
  CustomOperator,
  IdentifierAlt,
  Integer,
  TextFile,
  Dot,
  DotDot,
  Assignment,
  ElifKeyword,
  ElseKeyword,
  IfKeyword,
  StructKeyword,
  RArrow,
  EnumKeyword,
  MatchKeyword,
  RArrowAlt,
  VBar,
  ForeignKeyword,
  ModKeyword,
} from "./cst"
import { Diagnostics, UnexpectedCharDiagnostic } from "./diagnostics"
import { Stream, BufferedStream, assert } from "./util";

const EOF = '\uFFFF'

function isUpper(ch: string): boolean {
  return ch.toUpperCase() === ch;
}

function isWhiteSpace(ch: string): boolean {
  return /[\r\n\t ]/.test(ch);
}

function isIdentPart(ch: string): boolean {
  return /[a-zA-Z0-9_]/.test(ch);
}

function isIdentStart(ch: string): boolean {
  return /[a-zA-Z_]/.test(ch)
}

function isDecimalDigit(ch: string): boolean {
  return /[0-9]/.test(ch);
}

function toDecimal(ch: string): number {
  const code = ch.charCodeAt(0);
  assert(code >= 48 && code <= 57);
  return code - 48;
}

function isOperatorPart(ch: string): boolean {
  return /[+\-*\/%^&|$<>!?=]/.test(ch);
}

export class ScanError extends Error {

  public constructor(
    public file: TextFile,
    public position: TextPosition,
    public actual: string,
  ) {
    super(`Uncaught scanner error`);
  }

}

export class Scanner extends BufferedStream<Token> {

  private currLine = 1;
  private currColumn = 1;

  public constructor(
    public text: string,
    public textOffset: number = 0,
    public diagnostics: Diagnostics,
    private file: TextFile,
  ) {
    super();
  }

  private peekChar(offset = 1): string {
    const i = this.textOffset + offset - 1;
    return i < this.text.length ? this.text[i] : EOF;
  }

  private getChar(): string {
    const ch = this.textOffset < this.text.length
      ? this.text[this.textOffset++]
      : EOF
    if (ch === '\n') {
      this.currLine++;
      this.currColumn = 1;
    } else {
      this.currColumn++;
    }
    return ch;
  }

  private takeWhile(pred: (ch: string) => boolean): string {
    let out = ''
    for (;;) {
      const c0 = this.peekChar()
      if (!pred(c0)) {
        break;
      }
      this.getChar()
      out += c0;
    }
    return out;
  }

  private getCurrentPosition(): TextPosition {
    return new TextPosition(
      this.textOffset,
      this.currLine,
      this.currColumn
    );
  }

  public read(): Token {

    let c0: string;

    // Skip whitespace and comments
    for (;;) {

      for (;;) {
        c0 = this.peekChar();
        if (isWhiteSpace(c0)) {
          this.getChar();
          continue;
        }
        if (c0 === '#') {
          this.getChar();
          for (;;) {
            const c1 = this.getChar();
            if (c1 === '\n' || c1 === EOF) {
              break;
            }
          }
          continue;
        }

        // We failed to match a newline or line comment, so there's nothing to skip
        break;

      }

      const startPos = this.getCurrentPosition();
      this.getChar();

      switch (c0) {

        case '"':
        {
          const startPos = this.getCurrentPosition();
          let contents = '';
          let escaping = false;
          for (;;) {
            if (escaping) {
              const startPos = this.getCurrentPosition();
              const c1 = this.getChar();
              switch (c1) {
                case 'a': contents += '\a'; break;
                case 'b': contents += '\b'; break;
                case 'f': contents += '\f'; break;
                case 'n': contents += '\n'; break;
                case 'r': contents += '\r'; break;
                case 't': contents += '\t'; break;
                case 'v': contents += '\v'; break;
                case '0': contents += '\0'; break;
                case '\'': contents += '\''; break;
                case '\"': contents += '\"'; break;
                default:
                  throw new ScanError(this.file, startPos, c1);
              }
              escaping = false;
            } else {
              const c1 = this.getChar();
              if (c1 === '"') {
                break;
              } else {
                contents += c1;
              }
            }
          }
          return new StringLiteral(contents, startPos);
        }

        case EOF:
        {
          return new EndOfFile(startPos);
        }

        case '(': return new LParen(startPos);
        case ')': return new RParen(startPos);
        case '[': return new LBracket(startPos);
        case ']': return new RBracket(startPos);
        case '{': return new LBrace(startPos);
        case '}': return new RBrace(startPos);
        case ',': return new Comma(startPos);
        case ':': return new Colon(startPos);
        case '.': {
          const dots = c0 + this.takeWhile(ch => ch === '.');
          if (dots === '.') {
            return new Dot(startPos);
          } else if (dots === '..') {
            return new DotDot(startPos);
          } else {
            throw new ScanError(this.file, startPos, dots);
          }
        }

        case '+':
        case '-':
        case '*':
        case '/':
        case '%':
        case '&':
        case '^':
        case '|':
        case '$':
        case '<':
        case '>':
        case '=':
        case '!':
        case '?':
        {
          const text = c0 + this.takeWhile(isOperatorPart);
          if (text === '->') {
            return new RArrow(startPos);
          } else if (text === '=>') {
            return new RArrowAlt(startPos);
          } else if (text === '|') {
            return new VBar(startPos);
          } else if (text === '=') {
            return new Equals(startPos);
          } else if (text.endsWith('=') && text[text.length-2] !== '=') {
            return new Assignment(text, startPos);
          } else {
            return new CustomOperator(text, startPos);
          }
        }

        case '0':
        {
          const c1 = this.peekChar();
          switch (c1) {
            case 'x': // TODO
            case 'o': // TODO
            case 'b': // TODO
          }
        }
        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
        {
          let value = BigInt(toDecimal(c0));
          for (;;) {
            const c1 = this.peekChar();
            if (!isDecimalDigit(c1)) {
              break;
            }
            this.getChar();
            value = value * BigInt(10) + BigInt(toDecimal(c1));
          }
          return new Integer(value, 10, startPos);
        }

        case 'a':
        case 'b':
        case 'c':
        case 'd':
        case 'e':
        case 'f':
        case 'g':
        case 'h':
        case 'i':
        case 'j':
        case 'k':
        case 'l':
        case 'm':
        case 'n':
        case 'o':
        case 'p':
        case 'q':
        case 'r':
        case 's':
        case 't':
        case 'u':
        case 'v':
        case 'w':
        case 'x':
        case 'y':
        case 'z':
        case 'A':
        case 'B':
        case 'C':
        case 'D':
        case 'E':
        case 'F':
        case 'G':
        case 'H':
        case 'I':
        case 'J':
        case 'K':
        case 'L':
        case 'M':
        case 'N':
        case 'O':
        case 'P':
        case 'Q':
        case 'R':
        case 'S':
        case 'T':
        case 'U':
        case 'V':
        case 'W':
        case 'X':
        case 'Y':
        case 'Z':
        case '_':
        {
          const text = c0 + this.takeWhile(isIdentPart);
          switch (text) {
            case 'import': return new ImportKeyword(startPos);
            case 'pub': return new PubKeyword(startPos);
            case 'mut': return new MutKeyword(startPos);
            case 'let': return new LetKeyword(startPos);
            case 'import': return new ImportKeyword(startPos);
            case 'return': return new ReturnKeyword(startPos);
            case 'type': return new TypeKeyword(startPos);
            case 'if': return new IfKeyword(startPos);
            case 'else': return new ElseKeyword(startPos);
            case 'elif': return new ElifKeyword(startPos);
            case 'struct': return new StructKeyword(startPos);
            case 'enum': return new EnumKeyword(startPos);
            case 'match': return new MatchKeyword(startPos);
            case 'foreign': return new ForeignKeyword(startPos);
            case 'mod': return new ModKeyword(startPos);
            default:
              if (isUpper(text[0])) {
                return new IdentifierAlt(text, startPos);
              } else {
                return new Identifier(text, startPos);
              }
          }
        }

        default:

          // Nothing matched, so the current character is unrecognisable
          throw new ScanError(this.file, startPos, c0);
        }

    }

  }

}

const enum FrameType {
  Block,
  LineFold,
  Fallthrough,
}

const INIT_POS = new TextPosition(0, 0, 0);

export class Punctuator extends BufferedStream<Token> {

  private referencePositions: TextPosition[] = [ INIT_POS ];

  private frameTypes: FrameType[] = [ FrameType.Block ];

  public constructor(
    private tokens: Stream<Token>,
  ) {
    super();
  }

  public read(): Token {

    const t0 = this.tokens.peek(1);

    switch (t0.kind) {
      case SyntaxKind.LBrace:
        this.frameTypes.push(FrameType.Fallthrough);
        break;
      case SyntaxKind.EndOfFile:
      {
        if (this.frameTypes.length === 1) {
          return t0;
        }
        const frameType = this.frameTypes.pop()!;
        switch (frameType) {
          case FrameType.LineFold:
            return new LineFoldEnd(t0.getStartPosition());
          case FrameType.Block:
            return new BlockEnd(t0.getStartPosition());
        }
      }
    }

    const refPos = this.referencePositions[this.referencePositions.length-1];
    const frameType = this.frameTypes[this.frameTypes.length-1];

    switch (frameType) {

      case FrameType.Fallthrough:
      {
        if (t0.kind === SyntaxKind.RBrace) {
          this.frameTypes.pop()!;
        }
        this.tokens.get();
        return t0;
      }

      case FrameType.LineFold:
      {

        // This important check verifies we're still inside the line-fold. If
        // we aren't, we need to clean up the stack a bit and eventually return
        // a token that indicates the line-fold ended.
        if (t0.getStartLine() > refPos.line
          && t0.getStartColumn() <= refPos.column) {
          this.frameTypes.pop();
          this.referencePositions.pop();
          return new LineFoldEnd(t0.getStartPosition());
        }

        const t1 = this.tokens.peek(2);
        if (t0.kind === SyntaxKind.Dot && t0.getEndLine() < t1.getStartLine()) {
          this.tokens.get();
          this.frameTypes.push(FrameType.Block);
          return new BlockStart(t0.getStartPosition());
        }

        // If we got here, this is an ordinary token that is part of the
        // line-fold. Make sure to consume it and return it to the caller.
        this.tokens.get();
        return t0;
      }

      case FrameType.Block:
      {

        if (t0.getStartColumn() <= refPos.column) {

          // We only get here if the current token is less indented than the
          // current reference token. Pop the block indicator and leave the
          // reference position be for the edge case where the parent line-fold
          // continues after the block.
          this.frameTypes.pop();
          return new BlockEnd(t0.getStartPosition());

        }

        this.frameTypes.push(FrameType.LineFold);
        this.referencePositions.push(t0.getStartPosition());

        // In theory, we could explictly issue a LineFoldStart and let all
        // tokens be passed through in the FrameType.LineFold case. It does add
        // more logic to the parser for no real benefit, which is why it was
        // omitted.
        this.tokens.get();
        return t0;
      }

    }

  }

}

File: src/Scanner.cc

#include <unordered_map>

#include "bolt/Common.hpp"
#include "bolt/Text.hpp"
#include "bolt/Integer.hpp"
#include "bolt/CST.hpp"
#include "bolt/Diagnostics.hpp"
#include "bolt/DiagnosticEngine.hpp"
#include "bolt/Scanner.hpp"

namespace bolt {

static inline bool isWhiteSpace(Char Chr) {
  switch (Chr) {
    case ' ':
    case '\n':
    case '\r':
    case '\t':
      return true;
    default:
      return false;
  }
}

static inline bool isOperatorPart(Char Chr) {
  switch (Chr) { 
    case '+':
    case '-':
    case '*':
    case '/':
    case '^':
    case '&':
    case '|':
    case '%':
    case '$':
    case '!':
    case '?':
    case '>':
    case '<':
    case '=':
      return true;
    default:
      return false;
  }
}

static bool isDirectiveIdentifierStart(Char Chr) {
  return (Chr >= 65 && Chr <= 90) // Uppercase letter
      || (Chr >= 96 && Chr <= 122) // Lowercase letter
      || Chr == '_';
}

static bool isIdentifierPart(Char Chr) {
  return (Chr >= 65 && Chr <= 90) // Uppercase letter
      || (Chr >= 96 && Chr <= 122) // Lowercase letter
      || (Chr >= 48 && Chr <= 57) // Digit
      || Chr == '_';
}

static int toDigit(Char Chr) {
  ZEN_ASSERT(Chr >= 48 && Chr <= 57);
  return Chr - 48;
}

std::unordered_map<ByteString, NodeKind> Keywords = {
  { "class", NodeKind::ClassKeyword },
  { "do", NodeKind::DoKeyword },
  { "elif", NodeKind::ElifKeyword },
  { "else", NodeKind::ElseKeyword },
  { "enum", NodeKind::EnumKeyword },
  { "fn", NodeKind::FnKeyword },
  { "foreign", NodeKind::ForeignKeyword },
  { "if", NodeKind::IfKeyword },
  { "instance", NodeKind::InstanceKeyword },
  { "let", NodeKind::LetKeyword },
  { "match", NodeKind::MatchKeyword },
  { "mod", NodeKind::ModKeyword },
  { "mut", NodeKind::MutKeyword },
  { "pub", NodeKind::PubKeyword },
  { "return", NodeKind::ReturnKeyword },
  { "struct", NodeKind::StructKeyword },
  { "type", NodeKind::TypeKeyword },
};

Scanner::Scanner(DiagnosticEngine& DE, TextFile& File, Stream<Char>& Chars):
  DE(DE), File(File), Chars(Chars) {}

std::string Scanner::scanIdentifier() {
  auto Loc = getCurrentLoc();
  auto C0 = getChar();
  if (!isDirectiveIdentifierStart(C0)) {
    DE.add<UnexpectedStringDiagnostic>(File, Loc, std::string { C0 });
    return nullptr;
  }
  ByteString Text { static_cast<char>(C0) };
  for (;;) {
    auto C1 = peekChar();
    if (!isIdentifierPart(C1)) {
      break;
    }
    Text.push_back(C1);
    getChar();
  }
  return Text;
}

Token* Scanner::readNullable() {

  TextLoc StartLoc;
  Char C0;

  for (;;) {
    StartLoc = getCurrentLoc();
    C0 = getChar();
    if (isWhiteSpace(C0)) {
      continue;
    }
    if (C0 == '#') {
      auto C1 = peekChar(0);
      auto C2 = peekChar(1);
      if (C1 == '!' && C2 == '!') {
        getChar();
        getChar();
        auto Name = scanIdentifier();
        std::string Value;
        for (;;) {
          C0 = getChar();
          Value.push_back(C0);
          if (C0 == '\n' || C0 == EOF) {
            break;
          }
        }
        continue;
      }
      for (;;) {
        C0 = getChar();
        if (C0 == '\n' || C0 == EOF) {
          break;
        }
      }
      continue;
    }
    break;
  }

  switch (C0) {

    case static_cast<Char>(EOF):
      return new EndOfFile(StartLoc);

    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
    {
      Integer I = toDigit(C0);
      for (;;) {
        auto C1 = peekChar();
        switch (C1) {
          case '0':
          case '1':
          case '2':
          case '3':
          case '4':
          case '5':
          case '6':
          case '7':
          case '8':
          case '9':
            getChar();
            I = I * 10 + toDigit(C1);
            break;
          default:
            goto digit_finish;
        }
      }
digit_finish:
      return new IntegerLiteral(I, StartLoc);
    }

    case 'A':
    case 'B':
    case 'C':
    case 'D':
    case 'E':
    case 'F':
    case 'G':
    case 'H':
    case 'I':
    case 'J':
    case 'K':
    case 'L':
    case 'M':
    case 'N':
    case 'O':
    case 'P':
    case 'Q':
    case 'R':
    case 'S':
    case 'T':
    case 'U':
    case 'V':
    case 'W':
    case 'X':
    case 'Y':
    case 'Z':
    {
      ByteString Text { static_cast<char>(C0) };
      for (;;) {
        auto C1 = peekChar();
        if (!isIdentifierPart(C1)) {
          break;
        }
        Text.push_back(C1);
        getChar();
      }
      return new IdentifierAlt(Text, StartLoc);
    }

    case 'a':
    case 'b':
    case 'c':
    case 'd':
    case 'e':
    case 'f':
    case 'g':
    case 'h':
    case 'i':
    case 'j':
    case 'k':
    case 'l':
    case 'm':
    case 'n':
    case 'o':
    case 'p':
    case 'q':
    case 'r':
    case 's':
    case 't':
    case 'u':
    case 'v':
    case 'w':
    case 'x':
    case 'y':
    case 'z':
    case '_':
    {
      ByteString Text { static_cast<char>(C0) };
      for (;;) {
        auto C1 = peekChar();
        if (!isIdentifierPart(C1)) {
          break;
        }
        Text.push_back(C1);
        getChar();
      }
      auto Match = Keywords.find(Text);
      if (Match != Keywords.end()) {
        switch (Match->second) {
          case NodeKind::PubKeyword:
            return new PubKeyword(StartLoc);
          case NodeKind::LetKeyword:
            return new LetKeyword(StartLoc);
          case NodeKind::ForeignKeyword:
            return new ForeignKeyword(StartLoc);
          case NodeKind::MutKeyword:
            return new MutKeyword(StartLoc);
          case NodeKind::TypeKeyword:
            return new TypeKeyword(StartLoc);
          case NodeKind::ReturnKeyword:
            return new ReturnKeyword(StartLoc);
          case NodeKind::IfKeyword:
            return new IfKeyword(StartLoc);
          case NodeKind::ElifKeyword:
            return new ElifKeyword(StartLoc);
          case NodeKind::ElseKeyword:
            return new ElseKeyword(StartLoc);
          case NodeKind::MatchKeyword:
            return new MatchKeyword(StartLoc);
          case NodeKind::ClassKeyword:
            return new ClassKeyword(StartLoc);
          case NodeKind::InstanceKeyword:
            return new InstanceKeyword(StartLoc);
          case NodeKind::StructKeyword:
            return new StructKeyword(StartLoc);
          case NodeKind::EnumKeyword:
            return new EnumKeyword(StartLoc);
          case NodeKind::DoKeyword:
              return new DoKeyword(StartLoc);
          case NodeKind::FnKeyword:
              return new FnKeyword(StartLoc);
          default:
            ZEN_UNREACHABLE
        }
      }
      return new Identifier(Text, StartLoc);
    }

    case '"':
    {
      ByteString Text;
      bool Escaping = false;
      for (;;) {
        auto Loc = getCurrentLoc();
        auto C1 = getChar();
        if (Escaping) {
          switch (C1) {
            case 'a': Text.push_back('\a'); break;
            case 'b': Text.push_back('\b'); break;
            case 'f': Text.push_back('\f'); break;
            case 'n': Text.push_back('\n'); break;
            case 'r': Text.push_back('\r'); break;
            case 't': Text.push_back('\t'); break;
            case 'v': Text.push_back('\v'); break;
            case '0': Text.push_back('\0'); break;
            case '\'': Text.push_back('\''); break;
            case '"': Text.push_back('"'); break;
            default:
              DE.add<UnexpectedStringDiagnostic>(File, Loc, String { static_cast<char>(C1) });
              return nullptr;
          }
          Escaping = false;
        } else {
          switch (C1) {
            case '"':
              goto after_string_contents;
            case '\\':
              Escaping = true;
              break;
            default:
              Text.push_back(C1);
              break;
          }
        }
      }
after_string_contents:
      return new StringLiteral(Text, StartLoc);
    }

    case '\\':
      return new Backslash { getCurrentLoc() };

    case '.':
    {
      auto C1 = peekChar();
      if (C1 == '.') {
        getChar();
        auto C2 = peekChar();
        if (C2 == '.') {
          DE.add<UnexpectedStringDiagnostic>(File, getCurrentLoc(), String { static_cast<char>(C2) });
          return nullptr;
        }
        return new DotDot(StartLoc);
      }
      return new Dot(StartLoc);
    }

    case '+':
    case '-':
    case '*':
    case '/':
    case '^':
    case '&':
    case '|':
    case '%':
    case '$':
    case '!':
    case '?':
    case '>':
    case '<':
    case '=':
    {
      ByteString Text { static_cast<char>(C0) };
      for (;;) {
        auto C1 = peekChar();
        if (!isOperatorPart(C1)) {
          break;
        }
        Text.push_back(static_cast<char>(C1));
        getChar();
      }
      if (Text == "|") {
        return new VBar(StartLoc);
      } else if (Text == "->") {
        return new RArrow(StartLoc);
      } else if (Text == "=>") {
        return new RArrowAlt(StartLoc);
      } else if (Text == "=") {
        return new Equals(StartLoc);
      } else if (Text.back() == '=' && Text[Text.size()-2] != '=') {
        return new Assignment(Text.substr(0, Text.size()-1), StartLoc);
      }
      return new CustomOperator(Text, StartLoc);
    }

#define BOLT_SIMPLE_TOKEN(ch, name) case ch: return new name(StartLoc);

  BOLT_SIMPLE_TOKEN(',', Comma)
  BOLT_SIMPLE_TOKEN(':', Colon)
  BOLT_SIMPLE_TOKEN('(', LParen)
  BOLT_SIMPLE_TOKEN(')', RParen)
  BOLT_SIMPLE_TOKEN('[', LBracket)
  BOLT_SIMPLE_TOKEN(']', RBracket)
  BOLT_SIMPLE_TOKEN('{', LBrace)
  BOLT_SIMPLE_TOKEN('}', RBrace)
  BOLT_SIMPLE_TOKEN('~', Tilde)
  BOLT_SIMPLE_TOKEN('@', At)

  default:
    DE.add<UnexpectedStringDiagnostic>(File, StartLoc, String { static_cast<char>(C0) });
    return nullptr;

  }

}

Token* Scanner::read() {
  for (;;) {
    auto T0 = readNullable();
    if (T0) {
      // EndOFFile is guaranteed to be produced, so that ends the stream.
      return T0;
    }
  }
}

Punctuator::Punctuator(Stream<Token*>& Tokens):
  Tokens(Tokens) {
    Frames.push(FrameType::Block);
    Locations.push(TextLoc { 0, 0 });
  }

Token* Punctuator::read() {

  auto T0 = Tokens.peek();

  if (ShouldStartBlock) {
    ShouldStartBlock = false;
    Frames.push(FrameType::Block);
    return new BlockStart { T0->getStartLoc() };
  }

  switch (T0->getKind()) {
    case NodeKind::LBrace:
      Frames.push(FrameType::Fallthrough);
      break;
    case NodeKind::EndOfFile:
    {
      if (Frames.size() == 1) {
        return T0;
      }
      auto Frame = Frames.top();
      Frames.pop();
      switch (Frame) {
        case FrameType::Fallthrough:
          break;
        case FrameType::Block:
          return new BlockEnd(T0->getStartLoc());
        case FrameType::LineFold:
          return new LineFoldEnd(T0->getStartLoc());
      }
    }
    default:
      break;
  }

  auto RefLoc = Locations.top();
  switch (Frames.top()) {
    case FrameType::Fallthrough:
    {
      if (T0->getKind() == NodeKind::RBrace) {
        Frames.pop();
      }
      Tokens.get();
      return T0;
    }
    case FrameType::LineFold:
    {
      if (T0->getStartLine() > RefLoc.Line
        && T0->getStartColumn() <= RefLoc.Column) {
          Frames.pop();
          Locations.pop();
          return new LineFoldEnd(T0->getStartLoc());
      }
      if (isa<Dot>(T0)) {
        auto T1 = Tokens.peek(1);
        if (T1->getStartLine() > T0->getEndLine()) {
            Tokens.get();
            Frames.push(FrameType::Block);
            return new BlockStart(T0->getStartLoc());
        }
      }
      if (isa<DoKeyword>(T0)) {
        ShouldStartBlock = true;
      }
      return Tokens.get();
    }
    case FrameType::Block:
    {
      if (T0->getStartColumn() <= RefLoc.Column) {
        Frames.pop();
        return new BlockEnd(T0->getStartLoc());
      }

      Frames.push(FrameType::LineFold);
      Locations.push(T0->getStartLoc());

      return Tokens.get();
    }
  }

  ZEN_UNREACHABLE
}


}
